src/Symfony/Component/Security/Csrf/DoubleSubmitCsrfTokenManager.php

```diff
@@ -0,0 +1,185 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Csrf;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\Session\Session;
+use Symfony\Component\HttpKernel\Event\ResponseEvent;
+
+/**
+ * This CSRF token manager uses a combination of cookie and headers to validate non-persistent tokens.
+ *
+ * Double-Submit Validation: A JavaScript snippet on the client side is responsible for performing the
+ * double-submission. If the double-submit information is missing, we fall back to using the Origin or
+ * Referer headers.
+ *
+ * Fallback Scenarios: If neither double-submit nor Origin/Referer headers are available, it typically
+ * indicates that JavaScript is disabled on the client side (unless the JavaScript snippet was not
+ * properly implemented), or that the Origin header was not sent.
+ *
+ * By default, requests lacking both double-submit and origin information are deemed insecure.
+ *
+ * Security Consistency: When a session is found, a behavioral check is added to ensure that the
+ * validation method does not downgrade from double-submit to origin checks, or from origin checks to
+ * the accept fallback. This prevents attackers from exploiting potentially less secure validation
+ * methods once a more secure method has been confirmed as functional.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class DoubleSubmitCsrfTokenManager implements CsrfTokenManagerInterface
+{
+    public const HEADER_NAME = 'x-csrf-token';
+
+    public function __construct(
+        private RequestStack $requestStack,
+        private ?LoggerInterface $logger = null,
+        private string $headerName = self::HEADER_NAME,
+        private bool $acceptAsFallback = false,
+    ) {
+    }
+
+    public function getToken(string $tokenId): CsrfToken
+    {
+        return new CsrfToken($tokenId, $this->headerName);
```

Comment:
the `Csrftokenmanager` class uses random value to harden BREACH attacks. (https://github.com/symfony/symfony/commit/8b01095933e5073688b9bde2842974f33cc88412)


Maybe that logic should move outside the CsrfTokenManagers.
Also, I'm not sure randomizing the token is enough, we might need to randomize the length as well

Reply:
I don't think BREACH applies anymore: there's nothing stable to sniff, the random token is regenerated on the client-side for every single for submission.